File: clean-code-skills/configs/architecture/demo/python/app/application/price_order.py
"""Use case: depends inward only, and talks to the world through its own port."""
from __future__ import annotations

from app.application.ports import OrderRepository
from app.domain.money import Money


def price_order(items: list[int], repository: OrderRepository) -> Money:
    total = Money(0)
    for amount in items:
        total = total.plus(Money(amount))
    repository.save_total(total)
    return total

File: clean-code-skills/configs/architecture/demo/java/src/com/example/application/PriceOrder.java
package com.example.application;

import com.example.domain.Money;
import java.util.List;

/** Use case: depends inward only, and receives its collaborators as ports. */
public final class PriceOrder {

    private PriceOrder() {
    }

    public static Money totalOf(List<Integer> amounts, SaveTotals sink) {
        Money total = new Money(0);
        for (Integer amount : amounts) {
            total = total.plus(new Money(amount));
        }
        sink.save(total);
        return total;
    }

    /** The port belongs to the use case, not to the database package. */
    public interface SaveTotals {
        void save(Money money);
    }
}
